examples/ctc.dx

```diff
@@ -13,89 +13,123 @@ However, Dex's autodiff can produce the backward
 pass automatically.  That makes this code much shorter than
 most implementations.
 
-import stats  -- for LogSpace
+import stats  -- for log-space representation of probabilities.
 
-'### Helper functions
+'## Custom index sets
 
-IsBlank = Bool
+'### Fence and Posts
+Here we create a custom datatype and index set
+that interleaves the elements of a table with another set
+of values representing all the spaces in between those elements.
 
-def interleave {m v} (blank:v) (labels: m=>v) : (m & IsBlank)=>v =
-  -- Turns "text" into "t e x t " by first pairing each letter with a blank,
-  -- then flattening the pairs back into a single-index table.
-  pairs = for i j. case j of
-    True  -> blank
-    False -> labels.i
-  for (i, j). pairs.i.j
+data FenceAndPosts n =
+  Fence n
+  Posts (Post n)
 
-def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
-  -- Concatenates a single element to the beginning of a sequence.
-  for idx. case idx of
-    Left () -> first
-    Right i -> seq.i
+instance {a} [Ix a] Ix (FenceAndPosts a)
+  size = 2 * size a + 1
+  ordinal = \i. case i of
+    Fence j -> 2 * ordinal j + 1
+    Posts j -> 2 * ordinal j
+  unsafe_from_ordinal = \o.
+    case is_odd o of
+      True  -> Fence $ unsafe_from_ordinal _ (intdiv2 o)
+      False -> Posts $ unsafe_from_ordinal _ (intdiv2 o)
+
+instance {a} [NonEmpty a] NonEmpty (FenceAndPosts a)
```

Comment:
You don't need `NonEmpty a` here -- `FenceAndPosts` is always non-empty, because there is always at least one `Post`.

Reply:
Oh yeah, good point.

Reply:
Done.